[REEF-303] Clean up the construction of Transport implementation

taegeonum opened the pull request:
This contains the following things: 
- Make `NettyMessagingTransport` injectable 
- Move `TransportFactory` from `reef.io.network` to `reef.wake.transport`
- Refactor to use `TranportFactory` 

JIRA: [REEF-303](https://issues.apache.org/jira/browse/REEF-303)

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/GroupCommDriverImpl.java:
```diff
@@ -124,6 +126,19 @@ public GroupCommDriverImpl(final ConfigurationSerializer confSerializer,
+  /**
+   * @deprecated Have an instance injected instead.
+   */
+  @Deprecated
```
Why is this deprecated?

taegeonum replied:
We want to have an injected instance with Tang, instead of calling a constructor directly.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NetworkService.java:
```diff
@@ -125,18 +125,17 @@ public NetworkService(final IdentifierFactory factory,
-      final @Parameter(NetworkServiceParameters.NetworkServiceTransportFactory.class) TransportFactory tpFactory,
-      final @Parameter(NetworkServiceParameters.NetworkServiceHandler.class) EventHandler<Message<T>> recvHandler,
-      final @Parameter(NetworkServiceParameters.NetworkServiceExceptionHandler.class) EventHandler<Exception> exHandler) {
+      final IdentifierFactory factory,
```
Why this cleanup?

taegeonum replied:
because Tang does not inject this constructor.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/naming/NameClient.java:
```diff
@@ -72,6 +73,7 @@ public NameClient(final String serverAddr,
    * @param factory    an identifier factory
    * @param cache      a cache
    */
+  @Deprecated
```
Why is this deprecated?

bgchun replied:
Similar comments for other NameClient's.
Please double check that they're not needed or make sure Tang injects NameClient.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/naming/NameClient.java:
```diff
@@ -108,7 +110,9 @@ public NameClient(final String serverAddr,
    * @param cache      a cache
+   * @deprecated
    */
+  @Deprecated
```
Same question here. 
NameClient's not injected by Tang.

bgchun commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/naming/NameServerImpl.java:
```diff
@@ -100,7 +100,7 @@ public NameServerImpl(
       final int port,
       final IdentifierFactory factory,
       final ReefEventStateManager reefEventStateManager) {
-    this(port, factory, reefEventStateManager, LocalAddressProviderFactory.getInstance(), RangeTcpPortProvider.Default);
```
Where's TcpPortProvider?
We need to limit the range of TCP ports.

taegeonum replied:
TcpPortProvider is injected into TransportFactory.

bgchun commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/impl/DefaultRemoteManagerImplementation.java:
```diff
@@ -64,8 +64,7 @@
+  public static final String UNKNOWN_HOST_NAME = NettyMessagingTransport.UNKNOWN_HOST_NAME;
 
   /**
    * @deprecated have an instance injected instead. Or use RemoteManagerFactory.getInstance()
```
Do we need this constructor?

bgchun commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/impl/DefaultRemoteManagerImplementation.java:
```diff
@@ -89,8 +88,7 @@
+        new MessagingTransportFactory());
   }
 
   /**
```
Why is this deprecated? 
It's also injected by Tang.

The deprecated annotation seems to be over-used. Please go through them and delete ones for the constructors we will keep.

taegeonum replied:
We want to have an injected instance with Tang, instead of calling a constructor directly. @markusweimer Is this what you intended? But, I agreed that it is somewhat confusing which constructors are not needed anymore. Do we need another annotation for this?

bgchun commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/MessagingTransportFactory.java:
```diff
@@ -63,10 +64,10 @@ public MessagingTransportFactory() {
-                          final EventHandler<TransportEvent> clientHandler,
-                          final EventHandler<TransportEvent> serverHandler,
-                          final EventHandler<Exception> exHandler) {
+  public Transport getInstance(final int port,
```
getInstance is usually used when it returns a singleton object.
If the method returns a new object every time it gets called, newInstance would be a better name.

taegeonum replied:
I've changed `getInstance` to `newTransport`

bgchun commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/MessagingTransportFactory.java:
```diff
@@ -74,4 +75,31 @@ public Transport create(final int port,
+                               final int numberOfTries,
+                               final int retryTimeout,
+                               final TcpPortProvider tcpPortProvider) {
+    return new NettyMessagingTransport(hostAddress,
```
NettyMessagingTransport is not injectable yet?

bgchun commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java:
```diff
@@ -99,19 +109,51 @@
    * @param tcpPortProvider  gives an iterator that produces random tcp ports in a range
+   * @deprecated use the constructor that takes a LocalAddressProvider instead.
+   */
+  @Deprecated
```
Again, the same comment. All the constructors are deprecated...

bgchun commented:
@taegeonum I left some comments. There're a few confusing parts. Injection seems not to be fully done. Please address my comments and let me know when it's ready for another review.

bgchun commented:
@taegeonum Also, please clean up the Deprecated annotations.

taegeonum commented:
@bgchun I've addressed your comments. But, I think it would be good to create a separate issue for cleaning up constructors of `DefaultRemoteManager`.
